File: 523-continuous-subarray-sum/continuous-subarray-sum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */
var checkSubarraySum = function(nums, k) {
    const map = new Map()

    map.set(0, -1)
    let prefixSum = 0

    for (let i = 0; i < nums.length; i++) {
        prefixSum += nums[i]

        if (k !== 0) {
            prefixSum = prefixSum % k
        }

        if (map.has(prefixSum)) {
            if (i - map.get(prefixSum) >= 2) {
                return true
            } 
        } else {
            map.set(prefixSum , i)
        }
    }
    return false
};

File: 523-continuous-subarray-sum/continuous-subarray-sum.py
class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        prefix_hash = defaultdict(int)
        prefix_hash[0] = -1
        current_sum = 0

        for i, num in enumerate(nums):
            current_sum += num
            remainder = current_sum % k

            if remainder in prefix_hash and i - prefix_hash[remainder] >= 2:
                return True
            if remainder not in prefix_hash:
                prefix_hash[remainder] = i

        return False